Часть инвентаря

Rentgen993 opened the pull request:
- Добавлена нижняя строка инвентаря, хранящая предметы
- Добавлена вторая строка инвентаря, отвечающая за быстрое использование предметов
- Добавлена часть инвентаря, использующая артефакты-усилители-статов для игрока и врага

- Строки с лучом из первого умения выделены в отдельный класс

Commit:
Часть инвентаря

- Добавлена нижняя строка инвентаря, хранящая предметы
- Добавлена вторая строка инвентаря, отвечающая за быстрое использование предметов
- Добавлена часть инвентаря, использующая артефакты-усилители-статов для игрока и врага

- Строки с лучом из первого умения выделены в отдельный класс

FrostHoll reviewed (changes requested):
без лишних слов, поехали:
- **RaycastRay:**
Название класса никаким образом не отражает его содержимое. Так же класс не содержит никаких полей -> значит он должен быть статическим. Кроме того, он никак не участвует в жизненном цикле Unity -> не должен быть наследником MonoBehavior. Название методов, напр. `ability` не пишется с маленькой буквы, а в PascalCase. Название метода также не отражает его содержание. Способность чего?
- **Item:**
Если это часть артефакта, которую Аделя еще не успела доделать, то принимается. После появления кода эту часть переработать.
- **StatsManager:**
Это еще что за чудище? Почему у него ссылка на игрока и... одного врага? И он занимается добавлением предметов? Жесть... стереть этот код с лица земли, перенести функционал в абстрактный `Controller`. Учитывая то, что разницы между этими методами нет абсолютно, оправданность существования этого кода мне не понятна.
- **SecondLine:**
Еще раз я увижу использование KeyCode, буду бить по голове, долго и больно. Особенно когда оно очень много раз и в одном месте - ощущение просто отвратительное. Проверка литералов (`"Consumable"`) - это отвратительная вещь, одна неправильная буква, и код уже не работает. Переделать на перечисления.
- **OpenInventory:**
Зачем получать ссылку на Canvas, если самим канвасом (именно методами и свойствами самого канваса) мы не пользуемся? Мы пользуемся только включением и выключением компонента, что есть и у обычного GameObject, чем и нужно пользоваться.
- **slot:**
Хотя он и пустой, придерусь к его названию: Не называем классы с маленькой буквы, а в PascalCase!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
- **Inventory:**
`Items` не должен быть публичным.
- **Inventory и SecondLine:**
У них есть общие методы и поля. Необходимо выделить все в один суперкласс, и тогда кодовая база и того и другого сильно уменьшится.

**Итог:** пока очень неплохо, осталось дождаться Аделю и ее шикарный код (который я жду уже месяц) и тогда можно всерьез браться за все это дело. Рекомендуется ее поторопить и/или помочь.

Commit:
Часть инвентаря 2.0

- PaScAlCaSe СоБлЮдЁн
- RayCast переименован в CreatingDamageArea и стал статическим
- Чудище StatsManager уничтожено, строки кода перенесены в Controller
- KeyCode наверное переделан, не знаю, не понимаю
- Canvas переделан на GameObject
- Items теперь не публичный
- SecondLine перенесен в Inventory

FrostHoll reviewed (changes requested):
ну что, посмотрим, что поменяла Рената...

- **Inventory:**
Система InputSystem так и не введена. Просто добавлено поле InputAction для видимости, но нигде не использовано, чтение нажатий так и происходит через `Input.GetKeyUp`. Скрыться не получилось. Переделать.

- **OpenInventory:**
Умора. `GetComponent<GameObject>` вызывается у компонента, который и так является GameObject'ом. Достаточно было просто написать `this` или `gameObject`, если нужна ясность. И тут кстати тоже чтение нажатия через Input!

- **Controller:**
Видимо Ренату вообще никак не смущает код, который она написала. Вот вообще. Ее не колышет, что Controller - абстрактный класс, который наследуют И PlayerController И EnemyController, и что в сигнатуре ApplyEffect у артефакта стоит Controller. Какая разница, какая детализация контроллера, если конечному методу все равно? Еще и ссылки в поля получать... В итоге получается, что у контроллера игрока должна быть выставлена ссылка на самого себя... и врага еще к тому же...
Пожалуйста, сделай уже так:

```csharp
public void AddArtifact(Artifact artifact)
{
    // проверочка
    artifact.ApplyEffect(this);
}
```

И все, проблем никаких не будет, никаких полей тоже.

**Итог:**
Изучить как работает InputSystem в Unity и переделать чтение клавиш. Так же на бис перечитать 2.1, и узнать что означает ключевое слово `this`.

Commit:
Изменения

- инвентарь разделен 5 частей (сделала как отдельные классы, но потом поняла, что обратно все вместе запихнуть)
- inputAction фиг его знает, я не разобралась
- функционал перенесен в slot

FrostHoll committed:
Merge branch 'main' into Renata2

FrostHoll reviewed (approved):
Thanks for everything. We will patiently wait for your return.
Until next time.
